particles/Tutorial/Kotlin/Demo/TTTGame.kt

```diff
@@ -0,0 +1,167 @@
+            "message" to if (gs.gameOver == true) congratsMessage else "",
+            "playerDetails" to "$cpName playing as $cpAvatar"
+        )
+    }
+
+    override fun onHandleUpdate(handle: Handle) {
+        val gs = gameState.get() ?: TTTGame_GameState()
+        // Apply the moves
+        if (gs.gameOver != true) {
+            val board = gs.board ?: defaultGame.board!!
+            val boardList = board.split(",").toMutableList()
+            // Check the handle updated matches the current player
+            if (handle.name == "playerOneMove" && gs.currentPlayer == 0.0) {
+                applyMove(
+                    mv = playerOneMove.get()?.move?.toInt() ?: -1,
+                    avatar = playerOne.get()?.avatar ?: "",
+                    boardList = boardList,
+                    gs = gs
+                )
+            } else if (handle.name == "playerTwoMove" && gs.currentPlayer == 1.0) {
+                applyMove(
+                    mv = playerTwoMove.get()?.move?.toInt() ?: -1,
+                    avatar = playerTwo.get()?.avatar ?: "",
+                    boardList = boardList,
+                    gs = gs
+                )
+            }
+        }
+        if (hasReset()) {
+            gameState.set(defaultGame)
+            playerOneMove.set(TTTGame_PlayerOneMove())
+            playerTwoMove.set(TTTGame_PlayerTwoMove())
+            events.clear()
+        }
+        renderOutput()
+    }
+
+    override fun getTemplate(slotName: String): String = """
+            It is your turn <span>{{playerDetails}}</span>.
+            <div slotid="boardSlot"></div>
+            <div><span>{{message}}</span></div>
+            """.trimIndent()
+
+    private fun applyMove(
+        mv: Int,
+        avatar: String,
+        boardList: MutableList<String>,
+        gs: TTTGame_GameState
+    ) {
+        if (!mv.isValidMove(boardList)) return
+        boardList[mv] = avatar
+        gs.board = boardList.joinToString(",")
+
+        val cp = gs.currentPlayer ?: 0.0
+        gs.currentPlayer = (cp + 1) % 2
+        gameState.set(gs)
+
+        gameState.set(checkGameOver(boardList, gs, avatar))
+    }
+
+    private fun checkGameOver(
+        boardList: List<String>,
+        gs: TTTGame_GameState,
+        avatar: String
+    ): TTTGame_GameState {
+        // Check if the board is full, meaning the game is tied
+        gs.gameOver = !boardList.contains("")
+
+        // Check if the game is over
+        winningSequences.forEach { sequence ->
+            if (boardList[sequence[0]] != "" &&
+                boardList[sequence[0]] == boardList[sequence[1]] &&
+                boardList[sequence[0]] == boardList[sequence[2]]) {
+                gs.gameOver = true
+                gs.winnerAvatar = avatar
+            }
+        }
+        return gs
+    }
+
+    private fun hasReset() = events.any { it.type == "reset" }
+}
+
+private fun Int.isValidMove(boardList: List<String>) = this in 0..9 && boardList[this] == ""
```

Comment:
Unless this function would be used by other things in the same file, consider moving it inside the TTTGame class.